test/cpp/end2end/mock_test.cc

```diff
@@ -161,6 +162,86 @@ class FakeClient {
   EchoTestService::StubInterface* stub_;
 };
 
+class CallbackTestServiceImpl
+    : public EchoTestService::ExperimentalCallbackService {
+ public:
+  experimental::ServerUnaryReactor* Echo(
+      experimental::CallbackServerContext* context, const EchoRequest* request,
+      EchoResponse* response) override {
+    // Make the mock service return a choice of Status even though the
+    // real one doesn't, so that we can test the status options.
+    auto* reactor = context->DefaultReactor();
+    if (request->message().length() > 0) {
+      response->set_message(request->message());
+      reactor->Finish(Status::OK);
+    } else {
+      reactor->Finish(Status(StatusCode::INVALID_ARGUMENT, "Invalid request"));
+    }
+    return reactor;
+  }
+};
+
+class MockCallbackTest : public ::testing::Test {
+ protected:
+  CallbackTestServiceImpl service_;
+
+  ServerContext context_;
+};
+
+TEST_F(MockCallbackTest, MockedCallSucceedsWithWait) {
+  experimental::CallbackServerContext ctx;
+  EchoRequest req;
+  EchoResponse resp;
+  grpc::internal::Mutex mu;
+  grpc::internal::CondVar cv;
+  grpc::Status status;
+  bool status_set = false;
+  DefaultReactorTestPeer peer(&ctx, [&](::grpc::Status s) {
+    grpc::internal::MutexLock l(&mu);
+    status_set = true;
+    status = std::move(s);
+    cv.Signal();
+  });
+
+  req.set_message("mock 1");
+  auto* reactor = service_.Echo(&ctx, &req, &resp);
+  cv.WaitUntil(&mu, [&] {
+    grpc::internal::MutexLock l(&mu);
+    return status_set;
+  });
+  EXPECT_EQ(reactor, peer.reactor());
+  EXPECT_TRUE(peer.test_status_set());
+  EXPECT_TRUE(peer.test_status().ok());
+  EXPECT_TRUE(status_set);
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ(req.message(), resp.message());
+}
+
+TEST_F(MockCallbackTest, MockedCallSucceeds) {
+  experimental::CallbackServerContext ctx;
+  EchoRequest req;
+  EchoResponse resp;
+  DefaultReactorTestPeer peer(&ctx);
+
+  req.set_message("ha ha, consider yourself mocked.");
+  auto* reactor = service_.Echo(&ctx, &req, &resp);
+  EXPECT_EQ(reactor, peer.reactor());
```

Comment:
If the reactor we get from the service is the same as the reactor we get from the peer, what's the point in exposing it from the peer?

Reply:
This is a check that the service is actually using the default reactor. If the service isn't using the default reactor, the DefaultReactorTestPeer won't be able to be used.

Reply:
It would be better if this was automatically checked so that it fails if the user tries to mock a call to a service that doesn't use the default reactor. In the current state, if the user's service doesn't use the default reactor and they forget to add this EXPECT_EQ to their test, they might waste cycles debugging mysterious behavior. 

Since every call to DefaultReactorTestPeer is going to be followed by an invocation of a method on the service, why don't we let the DefaultReactorTestPeer invoke the user's method? Then it can always perform the pointer check. Something like:

DefaultReactorTestPeer peer(&ctx, &service_.Echo);
peer.invoke_method() // Invokes Echo and does the EXPECT_EQ(reactor, peer.reactor()) check

Reply:
Something like:

DefaultReactorTestPeer peer(&ctx);
peer.invoke_method(&service_.Echo, &req, &resp) // Invokes Echo and does the EXPECT_EQ(reactor, peer.reactor()) check

Reply:
I've filed #21228 to make sure that we do something about this, preferably as additive API.